cmd/juju/application/deployer/bundlehandler.go

```diff
@@ -425,7 +417,10 @@ func (h *bundleHandler) resolveCharmChannelAndRevision(charmURL string, charmBas
 		ch = ch.WithRevision(revision)
 	}
 
-	_, origin, err := h.constructChannelAndOrigin(ch, charmBase, charmChannel, cons)
+	// We return early with local charms, so here we know the charm must be from charmhub
+	// We also return early for charm urls with defined revisions, so use the revision from params here
```

Comment:
Please explain more, why would only some charm urls have a revision at this point?

Reply:
Turns out this wasn't quite the correct logic, since we only exit early if `ch.Revision >= 0 && charmChannel != ""`
So instead use `ch.Revision`, which we have defaulted to the parameter `revision` is it's -1

This actually removes the 'wtf' this line may have thrown up, so I don't think an explanatory comment is necessary anymore

Reply:
The code below still uses "revision" so needs to be changed to `ch.Revision`.

Reply:
Fixed now. This had been left off because I forgot to push an extra change